### PR TITLE
RSDK-2893 Implement NextPointCloud for replaypcd

### PR DESCRIPTION
**Pull request by jeremyrhyde**
**Purpose:**
This PR introduces the ability for a camera with model `replay_pcd` to connect to app and retrieve pointcloud data from it.  The data sent back to the camera is determined based on:

- The source camera given in the config
- The robot_id given in the config (optional)
- The timestamps given in the config (optional)
- The last data file returned by any previous calls to NextPointCloud

This data is then converted into a `pointcloud.PointCloud` before being returned.

In order to perform this action, a cloud connection is formed between the robot and app. This follows the pathway of CloudConnection -> rpc.ClientConn -> DataServiceClient

**Testing:**
Manual Testing was performed by uploading data from a rplidar at 1 sec intervals to app. Note: No other data had previously been uploaded from this robot

image

Replay cameras were then initialized with a variety of filters including:
- No source given (immediate error)
- No robot_id given (unexpected data, but correct, that had been generated months ago)
- No filter (all data returned before end of data error was seen)
- Filter with start time (data returned with time receives >= start time)
- Filter with end time (data returned with time receives < end time)
- Filter with start and end time (data returned with time receives >= start time and < end time)

Automated testing was also performed which required the bringing up of a new mockDataServiceServer that diverted requests to BinaryDataByFilter to its associated function defined in `replay_utils_test.go`. This allowed for automated testing of NextPointCloud by returning filtered artifacts. More details can be found in `replay_test.go` and `replay_utils_test.go`.

**JIRA Ticket:** 
https://viam.atlassian.net/browse/RSDK-2893

### CODE REVIEW

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (edaniels) [inline components/camera/replaypcd/replaypcd_utils_test.go:102]: you'll need to close the conn too so just wrap this in a bigger `func() error { ... }`

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (edaniels) [inline components/camera/replaypcd/replaypcd_utils_test.go:126]: let's add this to some internal/testutils/inject at this point since data manager builtin sync_test uses this

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (edaniels) [inline components/camera/replaypcd/replaypcd_utils_test.go:133]: I'd like to see a test written where this fails as well as BinaryDataByFilter so that we have good coverage around those fail points

**Review by edaniels — CHANGES_REQUESTED** [review]
nice! no codegen 8)

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd_utils_test.go:102]: I was thinking about doing just that, however, in the function `closeCloudConnection` in `replaypcd.go`, I close the provided cloudConn. And since. I call Close at the end of every run, this should get closed there.

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd_utils_test.go]: Added to `internal/testutils/inject` and updated tests in `replaypcd_utils_test.go` and `datamanager/builtin/sync_test.go`

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd_utils_test.go]: I have added a test, on line 52 that will force an error response from the call to AcquireConnection and check to make sure it was received. 

Tests for failure modes on BinaryDataByFilter have already been tested by the tests checking if errEndOfDataset was produced when no files were available to be read.

**Review by edaniels — APPROVED** [review]
LGTM for test stuff

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [inline components/camera/replaypcd/replaypcd.go:121]: We specified in the tech spec: "If no end time is specified, we will continue to poll for new data and the point cloud stream will not have an endpoint."

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [inline components/camera/replaypcd/replaypcd.go:166]: We must prevent uses of the pcdCamera after close.

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [inline components/camera/replaypcd/replaypcd.go:189]: I don't think this needs to be in a defer.

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [inline components/camera/replaypcd/replaypcd.go:172]: We need to prevent use of pcdCamera during reconfiguration, or if reconfiguration fails.

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [inline components/camera/replaypcd/replaypcd.go:206]: Can we avoid repeating this defer by deferring replay.closeCloudConnection(ctx) right after the connection is initialized, and then cancelling the deferred statement if the reconfiguration succeeds (before returning nil)?

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (tessavitabile) [inline components/camera/replaypcd/replaypcd_utils_test.go:118]: Remove commented code

**Review by tessavitabile — COMMENTED** [review]
Really nice work on this, especially on the testing! Added a few comments, but no need to wait for my sign-off before merging.

**Review by dmhilly — COMMENTED** [review]
Awesome work! I left an initial round of comments; I haven't had the chance to look at the tests yet so for now all my feedback is on the replay camera implementation.

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [reply components/camera/replaypcd/replaypcd.go:121]: Got it, that makes sense.

Comment (tessavitabile) [reply components/camera/replaypcd/replaypcd.go:121]: Can you please update this in the tech spec?

Comment (dmhilly) [reply components/camera/replaypcd/replaypcd.go:121]: Updated!

Comment (tessavitabile) [reply components/camera/replaypcd/replaypcd.go:121]: thanks!

File: components/camera/replaypcd/replaypcd_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_test.go:139]: These tests are great! I think we should also reduce code duplication here with table-driven tests, however. Lmk if you want to talk about what that would look like.

File: components/camera/replaypcd/replaypcd_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_test.go:254]: I think we should add tests showing that filtering on different values of source and robotID works too. Just to prove that our implementation of NextPointCloud is actually passing these values onto BinaryDataByFilter. Our implementation of BinaryDataByFilter could only accept a specific name/ID for source/robotID, for example, and return no results otherwise.

File: components/camera/replaypcd/replaypcd_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_test.go:273]: Can we add a test showing that the consumer can still get data from the replay camera after the replay camera has returned an errEndOfDataset? We can hack our implementation of BinaryDataByFilter to know when it's reach the end of its dataset, and then have an extra file (maybe just one of the existing files) that it returns after that point. This would simulate a new file being "synced," and if anyone ever tried to change the replay camera to just return errEndOfDataset indefinitely after it had returned no data, this test would fail, indicating to the engineer that they have broken an intended feature.

File: internal/testutils/inject/cloud_connection.go
Comment (dmhilly) [inline internal/testutils/inject/cloud_connection.go:20]:
```suggestion
// AcquireConnection returns a connection to the rpc server stored in the CloudConnectionService object.
```

File: internal/testutils/inject/cloud_connection.go
Comment (dmhilly) [inline internal/testutils/inject/cloud_connection.go:12]: Maybe something like this instead? It's not clear what this is mocking from the comment.

```suggestion
// CloudConnectionService is an implementation of the cloud.ConnectionService interface used for testing.
```

File: internal/testutils/inject/cloud_connection.go
Comment (dmhilly) [inline internal/testutils/inject/cloud_connection.go:28]:
```suggestion
// Close is used by the CloudConnectionService to complete the cloud.ConnectionService interface.
```

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_utils_test.go:31]:
```suggestion
// can be overwritten to allow developers to trigger desired behaviors during testing.
```

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_utils_test.go:147]:
```suggestion
	// Apply the time-based filter based on the seconds value in the start and end fields. Because artifacts
```

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_utils_test.go:149]: I didn't really understand how this worked until I looked at the code. I think a quick example in this comment would make this clearer. Something like:

```
// For example, if there are 15 files in the artifact directory, the start time is 2000-01-01T12:00:10Z 
// and the end time is 2000-01-01T12:00:14Z, we will return files 10-14.
```

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_utils_test.go:192]: This logic feels a bit unnecessarily complicated to me, with getting the range of files and then iterating over it. Can't we just do something like:

```
startTime := -1
endTime := -1
if startTime != nil { startTime = startTime.AsTime().Second() }
if endTime != nil { endTime = endTime.AsTime().Second() }
if startTime != -1 && startTime > numPCDFiles { return nil }
if endTime == -1 || endTime <= numPCDFiles { 
    if !last {
        return getFile(endTime)
    }
    if last >= numPCDFiles {
        return nil
    }
    return getFile(last)
}
```
This is pseudocode but hopefully makes sense - happy to sync up offline.

**Review by dmhilly — CHANGES_REQUESTED** [review]
More feedback, this time on the tests. Really nice coverage overall! Left a few comments about places where I feel we could test further and on strategies for de-duping some of the test code.

File: components/camera/replaypcd/replaypcd.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd.go:195]: I can do that, I originally thought that we should store these important values for future use cases where we might not want to only be able to get them via the filter field. But if that use case comes up we can add them back to the struct later.

File: components/camera/replaypcd/replaypcd.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd.go:206]: Defers cannot be cancelled except via a variable check inside it that would only perform said defer if set. https://stackoverflow.com/questions/61240777/how-to-cancel-a-deferred-statement

Would you like me to pursue this avenue? It is unlike to reduce the number of lines but would allow us to reuse the defer

File: components/camera/replaypcd/replaypcd.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd.go:172]: Can this be handled by the new mutex lock and `closed` variable introduced thanks to your presentation on Thursday?

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [reply components/camera/replaypcd/replaypcd.go:206]: I don't feel strongly. It would be better to have a single defer, so that we don't need to remember to do this at every return statement, but you don't need to put a lot of time into investigating this.

If you leave it as is (close the connection prior to each return statement), then the "defer" shouldn't be needed, since you can just close the connection immediately and then return.

File: components/camera/replaypcd/replaypcd.go
Comment (tessavitabile) [reply components/camera/replaypcd/replaypcd.go:172]: yes, thank would work!

File: components/camera/replaypcd/replaypcd_test.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd_test.go:317]: robot_id is not checked in config validation as it is entirely optional and since the return on New() is a `camera.Camera` we don't have access to that parameter

File: components/camera/replaypcd/replaypcd_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_test.go:245]: Would `i == numPCDFiles` be sufficient? The >= condition threw me off.

File: components/camera/replaypcd/replaypcd_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_test.go:226]: Great test! Can we just leave a comment at the top of this function describing what this test proves? I imagine this could be confusing to someone looking at this without context.

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (dmhilly) [inline components/camera/replaypcd/replaypcd_utils_test.go:142]: Nitpicky, but can we modify this comment to make it clear that this function just returns the index of the file?

**Review by dmhilly — CHANGES_REQUESTED** [review]
Awesome work!! 🎉 Just a couple final questions. Excited to get this change in.

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd_utils_test.go:135]: Code was copied from here along with comment: https://github.com/viamrobotics/rdk/blob/0f49904ec550e5a33fc649e3153b4d465256a02a/services/datamanager/builtin/capture_test.go#L260

File: components/camera/replaypcd/replaypcd_test.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd_test.go:245]: I think its best practice to make checks like these as robust as possible. Same reason in a for loop we do:

for(i = 0; i < size; i ++) {}

If anything else in the loop were to be added that might change the value of I we want to ensure failure as accessing an element (or in this case an artifact) above the defined size should/would cause an error

File: components/camera/replaypcd/replaypcd.go
Comment (jeremyrhyde) [reply components/camera/replaypcd/replaypcd.go:68]: Currently, we require the timestamps from the config to be in UTC to avoid this issue. we can consider also importing the local time, however, that could lead to issues on pis and linux systems which have poorly defined timezones. Keeping everything as UTC ensures consistency between systems but can introduce more work for the user.

Comment (dmhilly) [reply components/camera/replaypcd/replaypcd.go]: Thanks for checking! Spoke with Jeremy offline and as long as the values in the config are provided in UTC, this will work, since the data capture timestamps are in UTC as well. Jeremy did some manual testing on this as well.

File: components/camera/replaypcd/replaypcd_utils_test.go
Comment (dmhilly) [reply components/camera/replaypcd/replaypcd_utils_test.go:135]: Ok, that's fine then.

File: components/camera/replaypcd/replaypcd_test.go
Comment (dmhilly) [reply components/camera/replaypcd/replaypcd_test.go]: Spoke with Jeremy offline, I agree that safety is important for production code paths but since this is a test I'd prefer that we do the minimum required check here. It makes the test easier to grok and if someone modifies this code such that we can get `i > numPCDFiles`, I'd rather we hit an error getting the file so the developer is aware that they messed something up.

**Review by dmhilly — APPROVED** [review]
Awesome work!! Left one nit about a comment, but LGTM. No need to re-request approval 🙂

**Pull request merged**
